test/qdmi/test_driver.cpp

```diff
@@ -0,0 +1,379 @@
+        << "Failed to allocate session.";
+
+    ASSERT_EQ(QDMI_session_init(session), QDMI_SUCCESS)
+        << "Failed to initialize session.";
+
+    size_t devicesSize = 0;
+    ASSERT_EQ(QDMI_session_query_session_property(session,
+                                                  QDMI_SESSION_PROPERTY_DEVICES,
+                                                  0, nullptr, &devicesSize),
+              QDMI_SUCCESS)
+        << "Failed to retrieve number of devices.";
+    std::vector<QDMI_Device> devices(devicesSize / sizeof(QDMI_Device));
+    ASSERT_EQ(QDMI_session_query_session_property(
+                  session, QDMI_SESSION_PROPERTY_DEVICES, devicesSize,
+                  static_cast<void*>(devices.data()), nullptr),
+              QDMI_SUCCESS)
+        << "Failed to retrieve devices.";
+
+    for (auto* const dev : devices) {
+      size_t namesSize = 0;
+      ASSERT_EQ(QDMI_device_query_device_property(
+                    dev, QDMI_DEVICE_PROPERTY_NAME, 0, nullptr, &namesSize),
+                QDMI_SUCCESS)
+          << "Failed to retrieve the length of the device's name.";
+      std::string name(namesSize - 1, '\0');
+      ASSERT_EQ(
+          QDMI_device_query_device_property(dev, QDMI_DEVICE_PROPERTY_NAME,
+                                            namesSize, name.data(), nullptr),
+          QDMI_SUCCESS)
+          << "Failed to retrieve the device's name.";
+
+      ASSERT_FALSE(name.empty()) << "Device must provide a non-empty name.";
+
+      if (name == deviceName) {
+        device = dev;
+        break;
+      }
+    }
+  }
+
+  void TearDown() override {
+    QDMI_session_free(session);
+    finalize();
+  }
+};
+
+TEST_P(DriverTest, SessionSetParameterImplemented) {
+  EXPECT_EQ(QDMI_session_set_parameter(session, QDMI_SESSION_PARAMETER_MAX, 0,
+                                       nullptr),
+            QDMI_ERROR_INVALIDARGUMENT)
+      << "Devices must implement `QDMI_session_set_parameter`.";
+}
+
+TEST_P(DriverTest, JobCreateImplemented) {
+  QDMI_Job job = nullptr;
+  EXPECT_EQ(QDMI_device_create_job(device, &job), QDMI_ERROR_NOTSUPPORTED)
+      << "Devices must implement `QDMI_device_create_job`.";
+  QDMI_job_free(job);
+}
+
+TEST_P(DriverTest, JobSetParameterImplemented) {
+  QDMI_Job job = nullptr;
+  EXPECT_EQ(QDMI_job_set_parameter(job, QDMI_JOB_PARAMETER_MAX, 0, nullptr),
+            QDMI_ERROR_INVALIDARGUMENT)
+      << "Devices must implement `QDMI_job_set_parameter`.";
+}
+
+TEST_P(DriverTest, JobQueryPropertyImplemented) {
+  QDMI_Job job = nullptr;
+  EXPECT_EQ(
+      QDMI_job_query_property(job, QDMI_JOB_PROPERTY_MAX, 0, nullptr, nullptr),
+      QDMI_ERROR_INVALIDARGUMENT)
+      << "Devices must implement `QDMI_job_set_parameter`.";
+}
+
+TEST_P(DriverTest, JobSubmitImplemented) {
+  QDMI_Job job = nullptr;
+  EXPECT_EQ(QDMI_job_submit(job), QDMI_ERROR_INVALIDARGUMENT)
+      << "Devices must implement `QDMI_job_submit`.";
+}
```

Comment:
This test does not particularly test whether the function is implemented.
For that, a non-null job, would need to be passed.
This can still be a separate test that ensures the right failure conditions.
Same holds for the other job tests down below.